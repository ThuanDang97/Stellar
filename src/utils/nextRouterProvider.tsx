import { NextRouter } from 'next/router'
import { RouterContext } from 'next/dist/shared/lib/router-context'

const createMockRouter = (router?: Partial<NextRouter>): NextRouter => ({
  basePath: '',
  pathname: '/',
  route: '/',
  query: {},
  asPath: '/',
  back: jest.fn(),
  beforePopState: jest.fn(),
  prefetch: jest.fn(),
  push: jest.fn(),
  reload: jest.fn(),
  replace: jest.fn(),
  events: {
    on: jest.fn(),
    off: jest.fn(),
    emit: jest.fn(),
  },
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  defaultLocale: 'en',
  domainLocales: [],
  isPreview: false,
  ...router,
})

export const NextRouterProvider = ({
  children,
  router = { query: {} },
}: {
  children: React.ReactNode
  router: Partial<NextRouter>
}) => (
  <RouterContext.Provider value={createMockRouter(router)}>
    {children}
  </RouterContext.Provider>
)
