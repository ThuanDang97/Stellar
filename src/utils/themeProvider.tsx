import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
