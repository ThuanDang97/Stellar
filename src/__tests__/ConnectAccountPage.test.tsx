import { fireEvent, screen, waitFor } from '@testing-library/react'
import { NextRouterProvider } from '@utils/nextRouterProvider'
import Router from 'next/router'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

// Pages
import ConnectAccountPage from '@pages/connect-account'

// Constants
import {
  EMAIL_NOT_EXIST,
  INVALID_PASSWORD,
  PRIMARY_HEADER_URL,
} from '@constants/index'

// Mocks
import {
  APPLICATION_CONNECT_ACCOUNT,
  INPUT_VALUES,
  USER_ID,
} from '@mocks/mockData'

const query = { id: USER_ID }

describe('ConnectAccount Page', () => {
  beforeEach(() => {
    renderWithTheme(
      <NextRouterProvider router={{ query }}>
        <ConnectAccountPage />
      </NextRouterProvider>,
    )
  })

  it('should render connectAccountForm', () => {
    const connectAccountForm = screen.getByTestId('connectAccountForm')

    expect(connectAccountForm).toBeInTheDocument()
  })

  const setup = () => {
    const utils = renderWithTheme(
      <NextRouterProvider router={{ query }}>
        <ConnectAccountPage />
      </NextRouterProvider>,
    )
    const input = screen.getAllByTestId('iconInput') as HTMLInputElement[]
    return {
      input,
      ...utils,
    }
  }

  it('should change email field value and display the correct value', () => {
    const { input } = setup()

    fireEvent.change(input[0], { target: { value: INPUT_VALUES.text } })
    expect(input[0]).toHaveValue(INPUT_VALUES.text)
  })

  it('should change email field value and display the correct value', () => {
    const { input } = setup()

    fireEvent.change(input[1], { target: { value: INPUT_VALUES.password } })
    expect(input[1]).toHaveValue(INPUT_VALUES.password)
  })

  it('should change  email not exits and display error', () => {
    const { input } = setup()

    fireEvent.change(input[0], { target: { value: INPUT_VALUES.text } })
    const button = screen.getAllByTestId('button') as HTMLInputElement[]

    fireEvent.submit(button[0])
    waitFor(() => {
      expect(screen.getByTestId('errorMessage')).toHaveValue(EMAIL_NOT_EXIST)
    })
  })

  it('should change password not match incorrect and display error', () => {
    const { input } = setup()

    fireEvent.change(input[0], {
      target: { value: APPLICATION_CONNECT_ACCOUNT.email },
    })
    fireEvent.change(input[1], { target: { value: INPUT_VALUES.password } })
    const button = screen.getAllByTestId('button') as HTMLInputElement[]

    fireEvent.submit(button[0])
    waitFor(() => {
      expect(screen.getByTestId('errorMessage')).toHaveValue(INVALID_PASSWORD)
    })
  })

  it('should checkout new page when submit success', () => {
    const input = screen.getAllByTestId('iconInput') as HTMLInputElement[]
    fireEvent.change(input[0], {
      target: { value: APPLICATION_CONNECT_ACCOUNT.email },
    })
    fireEvent.change(input[1], {
      target: { value: APPLICATION_CONNECT_ACCOUNT.password },
    })

    const button = screen.getAllByTestId('button') as HTMLInputElement[]

    fireEvent.submit(button[0])

    waitFor(() => {
      expect(Router.push).toHaveBeenCalledWith(
        PRIMARY_HEADER_URL.CONFIRM_DETAIL.URL,
      )
    })
  })

  it('ConnectAccountPage matches snapshot', () => {
    const { asFragment } = renderWithTheme(
      <NextRouterProvider router={{ query }}>
        <ConnectAccountPage />
      </NextRouterProvider>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should switch to a new page when clicking back', () => {
    const button = screen.getAllByTestId('button')

    fireEvent.submit(button[0])

    waitFor(() => {
      expect(Router.push).toHaveBeenCalledWith(PRIMARY_HEADER_URL.DASHBOARD)
    })
  })
})
