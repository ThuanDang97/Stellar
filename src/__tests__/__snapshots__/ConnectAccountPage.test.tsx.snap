// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectAccount Page ConnectAccountPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ViewStyled-sc-rqe5sa-0 DashBoardPageStyled__WrapperMainContent-sc-2iw14z-1 dKdLBS ljMWJf"
    width="100%"
  >
    <div
      class="ViewStyled-sc-rqe5sa-0 cIuayL"
      display="flex"
    >
      <button
        class="ButtonStyled-sc-1ys9v52-0 lbfTyw"
        data-testid="button"
      >
        <span
          class="ButtonStyled__ButtonIconStyled-sc-1ys9v52-1 DjHyU"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 lFLaW"
            display="inline-block"
            height="20"
            width="20"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
        </span>
      </button>
    </div>
    <div
      class="ViewStyled-sc-rqe5sa-0 kBewKp"
      display="flex"
    >
      <h1
        class="TextStyled-sc-9xia41-0 TitleStyled-sc-t58pjn-0 TitleStyled__TitlePageStyled-sc-t58pjn-1 kTPsig ONSni ilgUUT"
        color="#242A30"
        font-family="AdobeCleanExtraBold"
        font-size="36"
        letter-spacing="2"
      >
        Connect Account
      </h1>
      <span
        class="TextStyled-sc-9xia41-0 jMvPcG"
        color="#4C5156"
        font-size="16"
        letter-spacing="0.9"
      >
        Enter your  credentials so we can pay this bill on your behalf. Don't worry, we do not store your login information.
      </span>
      <form
        class="ViewStyled-sc-rqe5sa-0 ViewStyled__BoxShadowStyled-sc-rqe5sa-2 fTVJlI hHmBMa"
        data-testid="connectAccountForm"
        novalidate=""
      >
        <div
          class="ViewStyled-sc-rqe5sa-0 cafaom"
          display="flex"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 gjRQva"
            cursor="pointer"
            height="80"
            width="80"
          >
            <span
              class="AvatarStyled-sc-1ppokaq-1 bOiqfJ"
              data-testid="avatarDefault"
            />
          </div>
          <span
            class="TextStyled-sc-9xia41-0 TitleStyled-sc-t58pjn-0 TitleStyled__TitlePopupStyled-sc-t58pjn-2 XkRb kyLpEc iyHIHp"
            color="#000"
            font-size="20"
            font-weight="700"
          >
             Account
          </span>
        </div>
        <div
          class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 gJtdND brcjID"
          color="#203F59"
          data-testid="inputIcon"
          display="block"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 fTsvHV kHphsF"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                alt="icon"
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <input
            class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
            data-testid="iconInput"
            name="email"
            placeholder="Email Address"
            type="email"
            value=""
          />
        </div>
        <div
          class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 kdxmlR brcjID"
          color="#203F59"
          data-testid="inputIcon"
          display="block"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 fTsvHV kHphsF"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                alt="icon"
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <input
            class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
            data-testid="iconInput"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <button
          class="ButtonStyled-sc-1ys9v52-0 gmKOVy"
          data-testid="button"
          disabled=""
          type="submit"
        >
          Connect  Account
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
